### PR TITLE
Enable controllers based on which APIs the master is serving.

### PR DESCRIPTION
**Pull request by brendandburns**
Ref: https://github.com/kubernetes/kubernetes/issues/15402

@lavalamp @kubernetes/goog-control-plane @davidopp

### CODE REVIEW

**Comment by lavalamp** [comment]
#14941 has merged, so you can rebase this on head now, I think.

**Comment by brendandburns** [comment]
@lavalamp comments addressed.

Now split into to commits. 
   * [first one](https://github.com/brendandburns/kubernetes/commit/4ebbabe475de749803be70c8973e82ed10692f30) adds a method for getting supported resources from the server
   * [second one](https://github.com/kubernetes/kubernetes/commit/9a35a6512fd889b60c1a340f0832307c5df1fd6b) uses that for namespace controller and controller enabling.

please take another look.

**Comment by brendandburns** [comment]
Rebased.

**Comment by lavalamp** [comment]
looking

**Comment by lavalamp** [comment]
mostly ok now, thanks. Excited about controller manager :)

**Comment by brendandburns** [comment]
Ok, comments addressed.  ptal.

thanks
--brendan

**Comment by brendandburns** [comment]
This should be (really) ready for review now...

**Comment by lavalamp** [comment]
OK, this code is kind of a disaster, but I believe it should work now. @caesarxuchao has some clean up planned.

**Comment by lavalamp** [comment]
LGTM if it passes tests.

**Comment by derekwaynecarr** [comment]
Awesome, after this PR lands I should be able to make a more generic pass
at namespace controller for all resource types to avoid having to write any
code when registering new namespaced resource types.

/cc @kubernetes/rh-cluster-infra

/cc @liggitt @deads2k this would be a cleaner pattern for how we disabled
stuff in atomic enterprise vs OpenShift

**Comment by brendandburns** [comment]
@lavalamp 

I realized of course, that this is dependent on the rename from 'experimental/v1alpha1' to 'extensions/v1beta1'

Let me know how you want to proceed.  We can either switch the strings to 'experimental/v1alpha1' and merge this, or we can wait for the rename PR to go in.  Your preference.

**Comment by lavalamp** [comment]
I think the rename PR is in-- just maybe not cherrypicked yet.

**Comment by brendandburns** [comment]
ok, I will try a rebase.  I also realized I needed to tweak the manifest files to remove the flag.

File: cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest [hunk 1]
```diff
@@ -12,9 +12,6 @@
 {% if pillar['allocate_node_cidrs'] is defined -%}
   {% set allocate_node_cidrs = "--allocate-node-cidrs=" + pillar['allocate_node_cidrs'] -%}
 {% endif -%}
-{% if pillar['enable_experimental_api'] is defined -%}
-  {% set enable_experimental = "--enable-experimental=" + pillar['enable_experimental_api'] -%}
-{% endif -%}
```
Comment (erictune) [inline cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest]: Is line 4 which says `{% set enable_experimental = "" -%}` dead code now?

Comment (erictune) [inline cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest]: If so, the pillar definition can be deleted too, in cluster/gce/configure-vm.sh?

Comment (mikedanese) [inline cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest]: Line 4 is definitely dead code but pillar definition is still used by apiserver.manifest I believe

Comment (lavalamp) [inline cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest]: Yeah, apiserver will still need to consume something like this, but it's now dead in controller manager.

Comment (erictune) [inline cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest]:
```console
$ $ grep enable_experimental[^_] -l -R cluster
cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest
```
Looks like just the one use currently.  And I think apiserver will determine what resources to enable from `runtime-config`, so we won't need the pillar there.

Comment (lavalamp) [inline cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest]: Yes, but we can remove that in a follow-up PR. I would like to get this merged ASAP.

**Comment by lavalamp** [comment]
@brendandburns can you rebase please?

**Comment by brendandburns** [comment]
Rebased, ptal.

--brendan

**Comment by lavalamp** [comment]
LGTM when tests pass

**Comment by mikedanese** [comment]
Didn't pickup shippable

**Comment by brendandburns** [comment]
there was a broken mesos test (why don't we build these as part of ./hack/test-go.sh?)

fixed now.

**Comment by lavalamp** [comment]
I think brendan is out so I will run the cherrypick script.

**Comment by lavalamp** [comment]
oh man this cherrypick sucks :(

**Comment by caesarxuchao** [comment]
Will it make sense to cherrypick #15659 (DiscoveryClient) directly?

**Comment by caesarxuchao** [comment]
I realized this PR includes some other fixes.

**Comment by lavalamp** [comment]
Cherry pick in #15736

**Comment by lavalamp** [comment]
@caesarxuchao-- Yeah, this has other changes that I'm really after.

**Pull request merged**